[Mono] Add Mono Profiler events into EventPipe.

lateralusX opened the pull request:
Add Mono profiler events into EventPipe opening up for ability to emit events directly from Mono profiler callbacks into a EventPipe provider. PR maps all existing Mono profiler callbacks to Mono specific EventPipe events in new native EventPipe provider, Microsoft-DotNETRuntimeMonoProfiler.

PR also prepares for GC events and some of them will work as is, but a couple of events fires from within GC STW and since EventPipe is not async safe, those needs additional work to be correctly handled, and are currently disabled, gc_event, gc_moves, gc_resize, gc_roots.

A couple of events needs to be enabled in Mono profiler to fire (even when enabled in EventPipe) like exception_clause and gc_allocation. Added a new env variable, MONO_DIAGNOSTICS that opens up ability to configure this. I added ability to add additional config into that variable, and you can also use it to set DOTNET_DiagnosticPorts variable as well, meaning you can control all setting MONO_DIAGNOSTICS env variable. For example:

MONO_DIAGNOSTICS=--diagnostic-ports=127.0.0.1:9000,connect,suspend

will be the same as setting DOTNET_DiagnosticPorts=127.0.0.1:9000,connect,suspend

MONO_DIAGNOSTICS=---diagnostic-mono-profiler=alloc,exception

will enable both mono profiler GC alloc as well as exception clause instrumentation (needs to be set before starting profiler).

And if you just want to use one env variable, you can combine above in MONO_DIAGNOSTICS:

MONO_DIAGNOSTICS=--diagnostic-ports=127.0.0.1:9000,connect,suspend ---diagnostic-mono-profiler=alloc,exception

lateralusX committed:
Add Mono Profiler events into EventPipe.

lateralusX commented:
//CC @josalem

lateralusX committed:
Drop callstack on exception_clause profile event.

Needed since instrumentation uses mono_profiler_raise_exception_clause
used without wrapper and that in turn will cause incomplete stacks due
to transitioning into native code without informing unwinder.

lateralusX committed:
Fix build error.

lateralusX committed:
Add support to configure MonoVM diagnostics using env variable.

Adding new MONO_DIAGNOSTICS env variable that can include diagnostic
specific configs as well as --diagnostic-ports, meaning that its possible
to use that instead of DOTNET_DiagnosticPorts variable. It also add
variable to set some mono profiler settings needed very early during
startup to get GC alloc as well as exception clause checks.

In order to set needed options early in process, EventPipe component
calls a specific component_init method setting up needed config.

lambdageek reviewed (approved):
Mono bits LGTM. Python code LGTM too, though I'm less familiar with it.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -416,2571 +457,4397 @@ profiler_exception_clause (
+		if (is_ok (error) && runtime_helpers) {
+			MonoMethodBuilder *method_builder = mono_mb_new (runtime_helpers, "CompileMethod", MONO_WRAPPER_RUNTIME_INVOKE);
+			if (method_builder) {
+				_ep_rt_mono_runtime_helper_compile_method = mono_mb_create_method (method_builder, method_signature, 1);
```
It will only build a MonoMethod instance, but shouldn't compile the method. It should also handle case where it fails, and in that case we won't emit the phantom frame in callstacks.

lateralusX commented:
@lambdageek Detected an issue with the enable/disable callback that I need to adjust.

lateralusX committed:
Fix enable/disable callback registrations based on enable/disable bit mask.

josalem reviewed (approved):
CC @brianrob since this is rather major change to the manifest. I'd like his input before we merge if possible.

The changes look good to me, but I can't really comment on the Mono end of things (looks like Aleksey already went through that though). That being said, I don't think manually reading through ~7k lines of changes will catch all issues. If possible, I'd like to see some tests added before we snap .net6 for RC.

lateralusX commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -72,11 +72,19 @@ def generateMethodSignatureWrite(eventName, template, extern, runtimeFlavor):
     return ''.join(sig_pieces)
 
+def includeProvider(providerName, runtimeFlavor):
+    if runtimeFlavor.coreclr and providerName == "Microsoft-DotNETRuntimeMonoProfiler":
```
It will only used at this place at the moment, so no need to make it more visible.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -416,2571 +457,4426 @@ profiler_exception_clause (
+	uint32_t reserved_flags = 0;
+	uint64_t binding_id = 0;
+
+	// Native methods are part of JIT table and already emitted.
+	// TODO: FireEtwMethodDCEndVerbose_V1_or_V2 for all native methods in module as well?
```
For future, this comment/code it not new in this PR, changes to the file moved the code around.

lateralusX commented:
> CC @brianrob since this is rather major change to the manifest. I'd like his input before we merge if possible.
> 
> The changes look good to me, but I can't really comment on the Mono end of things (looks like Aleksey already went through that though). That being said, I don't think manually reading through ~7k lines of changes will catch all issues. If possible, I'd like to see some tests added before we snap .net6 for RC.

The majority of changes is in ep-rt-mono.c and is mainly not new changes, its mainly GIT not being able to trace injections in the beginning of the file together with changes in the middle and then some extensions to the end. @lambdageek has already gone through the changes in that file.

We would like to get this in for P7 since we have some SDK teams interested in using it. I view this new provider as being internal or preview mode for .net6 and primarily for internal tooling.

Regarding tests, do we have anything that tests emitting specific events defined in ClrEtwAll.man from runtime? There are a lot of events defined in that manifest file, some are public, some are private and some are not used at all anymore, where do we have tests checking all of them when emitted from runtime?

I think we should take this in steps, land this for P7 in preview/private so we can get feedback from SDK and tools teams and then see if we should expose it further.

lateralusX commented:
The format of the events implemented in this PR closely mirror the format used by mono log profiler. Those events have already been available for many years in mono/mono, so this PR mainly makes sure we get the same information we already had in mono log profiler into EventPipe and nettrace.

lateralusX committed:
Add ability to specify callspec for method instrumentation.

Add --diagnostic-mono-profiler-callspec= to accept Mono callspec string.

Split keywords to enable method tracing and instrumentation, enables
ability to start instrumenting (but not emitting events) in one session
and then enable emitting events in later session.

If a callspec is used, instrumentation will be enabled on component init.

lateralusX commented:
@lambdageek added ability to set a callspec making use of enter/leave events more useful.

steveisok reviewed (approved):
I would prefer merging this asap and following up with other changes, if needed. It will help us better analyze and potentially improve MAUI startup times, so there is some urgency.

lateralusX committed:
Only init component once when using static component builds.

brianrob commented:
@josalem thanks for including me.  I reviewed the manifest changes, and they look good to me.  The manifest change itself is fairly self-contained, in that it doesn't touch any of the existing manifests, and just creates a new one, so for the most part, the risk is mostly contained to the Mono paths here.

One point worth calling out is that the keywords values in the new provider skip lots of potential values (e.g. it skips from 0x10 for Loader to 0x4000 to JIT.  This is all legal, and should work fine, but I wanted to point this out because compressing these down can make it easier in the future when adding keywords, so you don't have to look for holes in the existing values.  Totally your call if you want to address this or not.

lateralusX commented:
> @josalem thanks for including me. I reviewed the manifest changes, and they look good to me. The manifest change itself is fairly self-contained, in that it doesn't touch any of the existing manifests, and just creates a new one, so for the most part, the risk is mostly contained to the Mono paths here.
> 
> One point worth calling out is that the keywords values in the new provider skip lots of potential values (e.g. it skips from 0x10 for Loader to 0x4000 to JIT. This is all legal, and should work fine, but I wanted to point this out because compressing these down can make it easier in the future when adding keywords, so you don't have to look for holes in the existing values. Totally your call if you want to address this or not.

Mainly reused values from runtime provider where they overlapped, so there was an intention for the holes. Since this is a Mono specific provider, we can plug them as well going forward or reuse more of the runtime keywords that currently are defined where we have holes.

lambdageek reviewed (approved):
callspec and init changes lgtm

brianrob commented:
> Mainly reused values from runtime provider where they overlapped, so there was an intention for the holes. Since this is a Mono specific provider, we can plug them when needed going forward or reuse more of the runtime keywords that currently are defined where we have holes.

Sounds good.  FWIW, I'm not sure how much benefit we'll gain from trying to match these up, since the events themselves are all different, but then again, it won't hurt either.

lateralusX commented:
> > Mainly reused values from runtime provider where they overlapped, so there was an intention for the holes. Since this is a Mono specific provider, we can plug them when needed going forward or reuse more of the runtime keywords that currently are defined where we have holes.
> 
> Sounds good. FWIW, I'm not sure how much benefit we'll gain from trying to match these up, since the events themselves are all different, but then again, it won't hurt either.

When reused they are closely related to the kind/group of events emitted, JIT keyword will enable JIT events, Loader keyword will enable loader events, Thread keyword thread events etc.
